infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -74,118 +84,201 @@
     /**
-     * Perform a state transfer on a current node, if needed, and then
-     * propose a new layout based on a transfer result.
-     * If a state transfer was not needed, try merging the segments
-     * of a current layout and then proposing it.
-     * We utilize an exponential backoff since there can be cases
-     * when multiple nodes are proposing a new layout simultaneously.
+     * A data class that stores both the layout and the transferred segments.
+     */
+    @AllArgsConstructor
+    private static class LayoutTransferSegments {
+        @Getter
+        @NonNull
+        private final Layout layout;
+
+        @Getter
+        @NonNull
+        private final ImmutableList<TransferSegment> transferSegments;
+    }
+
+    /**
+     * Perform state transfer and return all the transferred segments along with the latest layout.
      *
-     * @param runtime         A corfu runtime.
-     * @param transferManager A transfer manager that runs the state transfer.
-     * @return A new layout, if a redundancy restoration occurred; a current layout otherwise.
+     * @param runtime         Current runtime.
+     * @param transferManager Transfer manager instance to perform a state transfer.
+     * @return Current layout and the list of transferred segments.
      */
-    @VisibleForTesting
-    Layout restoreWithBackOff(CorfuRuntime runtime, StateTransferManager transferManager)
+    private LayoutTransferSegments performStateTransfer(CorfuRuntime runtime,
+                                                        StateTransferManager transferManager)
             throws InterruptedException {
-
-        // Set up retry settings.
+        // Settings for the retries
         Consumer<ExponentialBackoffRetry> retrySettings = settings -> {
             settings.setBase(retryBase);
             settings.setExtraWait(extraWait.toMillis());
             settings.setBackoffDuration(backoffDuration);
             settings.setRandomPortion(randomPart);
         };
 
-        // Configure a number of retries.
-        // Atomic is used here to overcome a restriction on the mutation of a local variable
-        // within a lambda expression block.
         AtomicInteger retries = new AtomicInteger(restoreRetries);
         return IRetry.build(ExponentialBackoffRetry.class, RetryExhaustedException.class, () -> {
             try {
                 // Retrieve a current layout.
                 runtime.invalidateLayout();
                 Layout currentLayout = runtime.getLayoutView().getLayout();
-
+                if (currentLayout.getUnresponsiveServers().contains(currentNode)) {
+                    throw new IllegalStateException("Node is in the unresponsive list.");
+                }
                 log.info("State transfer on {}: Layout before transfer: {}",
                         currentNode, currentLayout);
+                // Retrieve a current cluster trim mark.
+                long trimMark = trimLog(runtime, currentLayout);
```

Comment:
Retrieve trim mark does not have retry and it will rely on the whole workflow's retry mechanism.
Since it will send getTrimMark request to all log unit severs, and TimeoutException/NetworkException will consume one time workflow retry. Is it a little dangerous?

Reply:
Sure I can add some retries.